### ISSUE TITLE
Decide where to define the "domain types" corresponding to the proto types

### ISSUE THREAD

**Issue by plafer** [0xPolygonMiden/miden-node#74]
See [discussion](https://github.com/0xPolygonMiden/miden-node/pull/55#discussion_r1385684223) and [comment](https://github.com/0xPolygonMiden/miden-node/pull/55#discussion_r1405368470)

**Comment by plafer** [comment]
## Structure
I suggest the following structure, based on how the ibc-rs repository handles it:

- `proto` crate: this crate contains the `.proto` files, as well as the generated versions. It does *not* contain conversions.
  - this should probably live in `miden-base`, or in a repository of its own
  - this is analogous to [ibc-proto-rs](https://github.com/cosmos/ibc-proto-rs/tree/main)
- Any type that has an analogous protobuf type defines the conversion to-from the protobuf type next to its definition

## ibc-rs example
In the ibc-rs ecosystem, we'll take the `ChannelEnd` type as an example. 
- The protobuf definition lives [here](https://github.com/cosmos/ibc-proto-rs/blob/9c8be3df748e9b7f9509e561d30d4ed5a3fc47e3/src/prost/ibc.core.channel.v1.rs#L11-L28) (this is the generated version, the protobuf definition actually lives in another repository, but we don't need to do that). We call this the "raw" type
- The "domain" type lives [here](https://github.com/cosmos/ibc-rs/blob/de2c530f941eaf993badce2fccc4f3cb352f95b8/ibc-core/ics04-channel/types/src/channel.rs#L105-L111), in its separate crate, which has the `proto` crate as a dependency. The conversion code lives with the domain type [here](https://github.com/cosmos/ibc-rs/blob/de2c530f941eaf993badce2fccc4f3cb352f95b8/ibc-core/ics04-channel/types/src/channel.rs#L125-L166)

## Benefits of proposed structure
I prefer this system over ours because
- The conversion code lives next to the domain type; everything is in one place. Contrast with our [conversion.rs](https://github.com/0xPolygonMiden/miden-node/blob/f4eefddb4d8e4c12ce8659f0a6ddc084482a269d/proto/src/conversion.rs) module which contains every and all conversions of all types. This makes it much harder to understand how each protobuf type relates to other "domain types" in our system.
- You can implement `TryFrom<Raw> for Domain` where the `Error` type lives in the domain crate
  - our current system prevents us from doing this. Specifically, `miden-node-proto` doesn't have a reference to `miden-block-producer`, and so cannot implement `TryFrom<Raw, Error = SomeErrorTypeInBlockProducerCrate>`

**Comment by bobbinth** [comment]
A few preliminary thoughts:

I think there are two classes of domain objects: (1) objects defined somewhere upstream (e.g., `RpoDigest` defined in `miden-crypto`) and (2) objects defined locally in this repo specifically for the purposes of Miden Node (e.g., `AccountInputRecord`).

I think for the first class, the conversions have to be in the same crate where the generated protobuf structs are (or at least I'm not sure if we have any other option).

For the second class, I think we have 3 options:
1. Put domain types into `proto` crate.
2. Put domain types into its own crate which depends on `proto` crate.
3. Put domain types into specific crates where they are used (e.g., `rpc`, `block_producer`).

Currently, I think we have a mix of 1 and 3. To take The example of `AccountInputRecord`:
- The generated type is named `AccountBlockInputRecord` and is defined in the `proto` crate [here](https://github.com/0xPolygonMiden/miden-node/blob/main/proto/src/generated/responses.rs#L88C12-L88C35).
- A domain type called `AccountInputRecord` is defined in the `proto` crate [here](https://github.com/0xPolygonMiden/miden-node/blob/main/proto/src/domain/mod.rs#L5). This type is used in the `block_producer` crate.
- Another domain type called `AccountStateWithProof` is defined in the `store` crate [here](https://github.com/0xPolygonMiden/miden-node/blob/main/store/src/state.rs#L73C12-L73C33). This type is used by the store.

I think this situation is not ideal - we should probably pick one approach and stick with it.

In my opinion, options 1 and 2 are fairly close to each other as both of them mean we have a centralized location for defining domain types. For example, we could still keep domain types in the `proto` crate but improve on the current structure by moving conversion code close to the struct definitions. Specifically, instead of having one big `conversions.rs` file, we could put conversion code for `AccountInputRecord` right next to this struct.

So, the main question for me is whether we want to define domain structs in a central place (options 1/2) or have them defined in places where they are used (option 3).

The obvious downside of going with option 3 is that it could lead to code duplication. The potential upside is that it is more flexible (requests/responses could be built in different ways in different places).

**Comment by bobbinth** [comment]
One other thing to consider which is somewhat related to this (though, maybe should be its own issue/discussion) - we currently store protobuf-serialized objects in the store's database.

The main benefit is extra efficiency, but it does mean that we treat protobuf types as domain types - which feels odd. So, the question is - should we take a performance hit and translate protobuf objects to domain types before using them in meaningful ways in node components?

**Comment by plafer** [comment]
I see the domain types really as "really close" to the protobuf version; they only exist to address quirks from protobuf,such as every non-native type being optionally present, but we actually expect them all to be present always. Hence, I don't think we'd need the flexibility that option 3 gives; IMO, there's always only one conversion that "makes sense".

I'm indifferent whether we go with option 1 or 2, assuming we don't dump everything in a single `conversions.rs`. I'd probably lean towards option 2 though, since it provides a crate-level separation between both kinds of definitions (enabling things like functions accessible to the entire crate that are only useful for either proto or domain type stuff).

**Comment by bobbinth** [comment]
If we go with option 1, I wonder if it makes sense to rename the crate to something like `miden-node-core` and merge it with `utils` crate. The structure of the new crate could be something like:
```
├── proto
├── src
│   ├── domain
│   ├── generated
│   ├── utils
│   │   ├── config
│   │   ├── logging
│   │   └── hex.rs
│   ├── errors.rs
│   └── lib.rs
├── build.rs
└── Cargo.toml
```

**Comment by hackaugusto** [comment]
IMO we already have the domains clearly defined:

1. The domain of the servers is the networking, since we use protobuf, the generate code itself are the domain structs.
2. The code the servers are driving, like the VM and the protocol, have their own domain already, and are on their respective repos like `crypto`,  `miden-vm`, and `miden-objects`.

So a struct like `BlockWitness` should have gone to `miden-objects` and there should be a conversion in the protobuf to it. The conversions will always be on the protobuf crate because of the orphan rules that bobbin mentioned.

Another way of saying the above is that `protobuf` should not have a `domain` mod.

**Comment by plafer** [comment]
After the call we just had, my opinion is that the repo structure should look like:


```
     ┌────────────────────────────── miden-client
     │                                  │
     │                                  │
     ▼                                  │
miden-proto ◄─────── miden-base◄────────┤
     ▲                                  │
     │                                  │
     │                                  │
     └────────────────────────────── miden-node
```

- `miden-proto`: contains all the proto types that need to be sent over the wire
  - It is necessary to put all proto definitions in the same place because they have dependencies amongst them (e.g. `AccountBlockInputRecord` in `miden-node` depends on `AccountId` in `miden-base`)
- `miden-base`: contains the domain types that are shared by both the node and client (and possibly other consumers) - e.g. `AccountId`
  - Note: I think we're treating `AccountId` wrong currently by [redefining it in the node](https://github.com/0xPolygonMiden/miden-node/blob/4d626c341aedb9529db320912331f552b1ab2087/store/src/types.rs#L2). The node should always use `AccountId` in `miden-base`, and `miden-base` should define how it is converted to its protobuf equivalent.
- `miden-node` and `miden-client`: define the domain types that only they need (e.g. [`AccountInputRecord`](https://github.com/0xPolygonMiden/miden-node/blob/4d626c341aedb9529db320912331f552b1ab2087/proto/src/domain/mod.rs#L5)).
  - I'm not sure if `miden-client` needs to define its own types, but if ever it does, this structure would support it

One thing I had forgotten to consider previously is that some domain types live in `miden-base`. Therefore, "option 1" doesn't work anymore, because `miden-base` and `miden-client` also need access to the protobuf definition to provide the conversion to the corresponding domain type (which they would define - e.g. `AccountId` for `miden-base`).

A few notable points:

- All conversions then need to be in either `miden-base`, `miden-node` and `miden-client` (next to the corresponding domain type).
  - They can't be in the `miden-proto` crate because the `miden-proto` crate doesn't have access to the domain type definitions anymore.
- One big change from our current structure is that `miden-base` would now be aware that some of its types need to be convertible to a corresponding protobuf version.
  - I think this makes sense though - the consumers of `miden-base` need to use protobuf, and so `miden-base` should provide that capability
- @hackaugusto thinks that it will be too painful to make updates that span multiple repos. This is a fair point. The downside of our current approach though is that it is confusing even to us where things should be defined. I'm not sure which is the lesser evil.

**Comment by bobbinth** [comment]
I would actually go into the opposite direction and do something like this (which is basically what we have now):
```
     ┌────────────────────────────── miden-client
     │                                  │
     │                                  │
     ▼                                  │
miden-base ◄─────── miden-proto◄────────┤
     ▲                                  │
     │                                  │
     │                                  │
     └────────────────────────────── miden-node
```
The main reason for this is that I don't think of protobuf definitions as being a part of the core protocol. That is, if, in the future, we decide to ditch protobuf for anything else, `miden-base` shouldn't be affected.

I believe that with this structure we can still achieve most of the goals you mentioned. Specifically:
- `miden-base`: contains the domain types that are shared by both the node and client (and possibly other consumers).
- `miden-proto`:  contains all the proto types that need to be sent over the wire.
  - All conversions between proto types and domain types live here as well.
- `miden-node` and `miden-client`: define the domain types that only they need.

One question is where `miden-proto` crate should live. I think for now it is fine to keep it in this repo. In the future, we could actually move it into `miden-base` repo or create a separate repo for it altogether - but I think now it would create too much inconvenience.

**Comment by plafer** [comment]
> That is, if, in the future, we decide to ditch protobuf for anything else, miden-base shouldn't be affected.

Agreed.

One question remaining for me in this case is: where to put the protobuf definitions of the domain types defined in `miden-node` and `miden-client`? `miden-proto` defines the conversions with `miden-base` domain types, but it can't do that with `miden-node` and `miden-client`. It seems like for `miden-node/client` we'd need to put the protobuf definitions in `miden-proto`, and conversions in `miden-node/client`. So `miden-proto` would contain some conversions, but not all. Which might be ok?

**Comment by hackaugusto** [comment]
@bobbinth what is the approach you want me to take for this?

**Comment by bobbinth** [comment]
I think there are several things here, and I think we can do them sequentially (but also maybe makes sense to combine some of them):

1. Move away from storing protobuf messages in the database. This implies that before storing data in the DB (e.g., account ID, digest, etc.), we convert such data to our "native" types (e.g., types defined in or re-exported from `miden-base`).
2. Move conversions between protobuf messages and native types into `proto/src/domain` module. Here, instead of having a single large module, we should split the code up into sub-modules - one per domain type. For example: `proto/src/domain/account_id.rs`, `proto/src/domain/digest.rs`, `proto/src/domain/block_header.rs` etc.
3. For protobuf messages which do not map to native domain types from `miden-base`, we define "sister" types which remove `Option` from relevant fields. Let's take `AccountUpdate` as an example:

Currently, the generated Rust struct for this message is defined roughly as follows:
```Rust
pub struct AccountUpdate {
    pub account_id: ::core::option::Option<super::account::AccountId>,
    pub account_hash: ::core::option::Option<super::digest::Digest>,
}
```
Where `accountId` and `Digest` are protobuf messages. The "sister" type for this would be defined as:
```Rust
pub struct AccountUpdate {
    pub account_id: AccountId,
    pub account_hash: Digest,
}
```
Where `AccountId` and `Digest` are structs defined in `miden-base` (and `Digest` is actually `RpoDigest` defined in `miden-crypto` and re-exported from `miden-base`).

This sister type would be defined in the `src/proto/domain` module and conversions to/from protobuf message type would be defined there as well.

Note that we actually already have a "sister" type for `AccountUpdate` defined in the `store` called `AccountState`. This type would go away once we define the new `AccountUpdate` type.

4. Once the above is done, in most cases, response handlers would convert protobuf message types into corresponding domain "sister" types early on, and then rest of the code would deal only with the domain types.
5. As a part of the above, we may also want to normalize some protobuf message names.
6. Lastly, and this could be a part of a separate workstream altogether, we could consider merging `proto` and `utils` crates into something like `miden-node-core` (see [this](https://github.com/0xPolygonMiden/miden-node/issues/74#issuecomment-1867411314) comment).

All of the above would probably be done in several PRs and it is also possible that I missed something.

**Comment by Dominik1999** [comment]
Is this issue likely to be close this week?

**Comment by polydez** [comment]
@Dominik1999 
> Is this issue likely to be close this week?

Yes, I'm going to finish work on it today or tomorrow, hope, my changes will be approved.
